#560 itemlist headers

joncppl opened the pull request:
- Allow configurable column headers are configured in itemlist. 
- configure all pages using itemlist to have appropriate headers.
- (events will need to be updated to have headers following merge)
- The Article and Page index pages weren't using the reusable itemlist component. Now they do.

joncppl committed:
implement itemlist column headers

joncppl committed:
update Article and Page pages to use ItemIndexPage

joncppl committed:
column header font weight

psiemens committed:
Merge branch 'develop' into #560-itemlist-headers

psiemens committed:
Update styles

psiemens commented on dispatch/static/manager/src/js/components/ItemList/ItemListColumnHeaders.js:
```diff
@@ -0,0 +1,25 @@
+  }
+
+  // blank first column for checkbox
+  const headers = R.insert(0, '', props.headers)
```
In this case i think it might be cleaner to just write:

```js
const headers = [''] + props.headers
```

joncppl replied:
I believe `const headers = [''] + props.headers` would downcast the operands into strings before concatenation, so I don't think it'd work.

Instead I'll just have the first pseudoelement be hardcoded in jsx

psiemens commented on dispatch/static/manager/src/js/pages/ItemIndexPage.js:
```diff
@@ -83,8 +84,20 @@ export default class ListItemsPageComponent extends React.Component {
-    const titleString = this.props.typePlural.replace(/^\w/, m => m.toUpperCase())
+    const titleString = this.props.pageTitle
+      // Make the title start with uppercase letter
+      || this.props.typePlural.replace(/^\w/, m => m.toUpperCase())
```
Not sure if I love this approach, just because it's hidden from whoever is using the component. I think I'd prefer to make the `pageTitle` prop required

psiemens commented on dispatch/static/manager/src/js/pages/Articles/ArticleIndexPage.js:
```diff
@@ -1,190 +1,72 @@
-
-  componentDidUpdate(prevProps) {
-
-    if (this.isNewSection(prevProps, this.props) || this.isNewQuery(prevProps, this.props)) {
```
I *really* like the abstraction provided by `ItemIndexPage`, but in this case we lose the `isNewSection` check, which prevents the user from switching between different section tabs. Can we maybe add an optional prop to `ItemIndexPage` that allows us to extend this function?

psiemens reviewed (changes requested):
Awesome stuff -- definitely a much-needed enhancement. I pushed a commit with some style tweaks and added a few comments to the code

joncppl committed:
fix articles listing in sections

Add to ListItemPage,the functions extraRelistCondition and extraQueryHandler as props, used by ArticleIndexPage to add the section to the query object and to refresh the list when the section changes.

Also pageTitle is a required prop

psiemens commented on dispatch/static/manager/src/js/components/ItemList/ItemListColumnHeaders.js:
```diff
@@ -1,24 +1,21 @@
 
   return (
     <li className='c-item-list__item c-item-list__column-headers'>
+      <div key={0} className='c-item-list__item__cell' />
```
Nice -- I like this solution

psiemens commented on dispatch/static/manager/src/js/pages/Articles/ArticleIndexPage.js:
```diff
@@ -58,7 +58,15 @@ function ArticlePageComponent(props) {
         item => item.latest_version + ' revisions'
       ]}
-
+      extraRelistCondition={(prevProps, props) => {
```
I think it makes more sense to name this something like `shouldReload`

psiemens commented on dispatch/static/manager/src/js/pages/Articles/ArticleIndexPage.js:
```diff
@@ -58,7 +58,15 @@ function ArticlePageComponent(props) {
+      extraRelistCondition={(prevProps, props) => {
+        return prevProps.location.query.section !== props.location.query.section
+      }}
+      extraQueryHandler={(query, props) => {
```
Rename to `queryHandler` -- don't think extra needs to be included

psiemens commented on dispatch/static/manager/src/js/pages/ItemIndexPage.js:
```diff
@@ -24,6 +25,10 @@ export default class ListItemsPageComponent extends React.Component {
       query.listItem = this.props.location.query.listItem
     }
 
+    if (typeof this.props.extraQueryHandler === 'function') {
```
Instead of checking the type here, it would be better to add these props to `propTypes` and then just do `if (this.props.extraQueryHandler)`

Alternatively, you could use `defaultProps` and a stub function like so to avoid the conditional entirely:

```js

ListItemsPageComponent.defaultProps = {
  extraQueryHandler: (query) => query
}
```

psiemens reviewed (changes requested):
Just a few suggestions -- looks good otherwise

joncppl committed:
change method names and prop organization
